### ISSUE TITLE
"Load Order / Name Resolution Problem" with ruby core class mixins

### ISSUE THREAD

**Issue by jcavalieri** [lsegal/yard#1353]
I'm getting a *Load Order / Name Resolution Problem* when having a mixing on a core ruby object.

The error message mentions changing the load order. I don't know what that means. I did try `bundle exec yard doc ./lib/ext/**/*.rb ./lib/app.rb`, but that failed in the same way.

Is this a config issue on my part? Or could it be a bug?

## Steps to reproduce

See https://github.com/jcavalieri/yard_issue/ for a small example where it fails.

## Actual Output

```
$ bundle exec yard
[warn]: Load Order / Name Resolution Problem on Hash:
-
Something is trying to call mixins on object Hash before it has been recognized.
This error usually means that you need to modify the order in which you parse files
so that Hash is parsed before methods or other objects attempt to access it.
-
YARD will recover from this error and continue to parse but you *may* have problems
with your generated documentation. You should probably fix this.
-

[error]: Unhandled exception in YARD::Handlers::Ruby::MixinHandler:
  in `lib/app.rb`:7:

	7: Hash.include Ext::Core::Hash::Filtering

Files:           2
Modules:         4 (    3 undocumented)
Classes:         1 (    0 undocumented)
Constants:       0 (    0 undocumented)
Attributes:      0 (    0 undocumented)
Methods:         2 (    0 undocumented)
 57.14% documented
```

## Expected Output

Not to have an issue with mixins of core classes.

## Environment details:

- OS: OSX latest
- Ruby version: ruby-2.5.8
- YARD version: yard (~> 0.9.25)

**Comment by jcavalieri** [comment]
I've done a bit of sleuthing. I wonder if it might be related to this commit:

https://github.com/lsegal/yard/commit/81d226099d7615859a1222420b9f155c1e8dbbd7

I down graded to version 0.9.20 and it is working now. So I believe this was recently introduced.

**Comment by lsegal** [comment]
This is very likely going to be considered expected behavior. There is nothing incorrect about the behavior you are seeing: YARD is seeing a class being mutated that it knows nothing about; it is letting you know that this include call is being discarded due to the fact that the related Hash class is not being included in your documentation.

Note that this is a warning, not an error, and while verbose, it's something you can ignore if you do not want Hash documented. The easy workaround here would be to open the Hash class before sending include and marking it `@api private` (with `--no-api private` in your `.yardopts`):

```ruby
# @api private
class Hash
  include Ext::Core::Hash::Filtering
end
```

The reason this is failing as of 0.9.20 is because support for the `Const.include` syntax was added in that version to explicitly support this formatting; these type of inclusions for known classes had been previously ignored.

**Comment by jcavalieri** [comment]
Hi Isegal,

Thanks for the response and the explanation.

Maybe I'm missing something, but wouldn't this ...

```
# @api private
class Hash
  include Ext::Core::Hash::Filtering
end
```

... include the mixin globally to the application, instead of just when just needed? We have it as a dev practice to not do global mixins.

Also, I was surprised to hear it was just a warning, I see now that our CI build has `--fail-on-warning` set. Which I suspect we still want to keep.

By chance, is there an inline mechanism like you showed above with `# @api private`?

**Comment by lsegal** [comment]
> but wouldn't this include the mixin globally to the application, instead of just when just needed? We have it as a dev practice to not do global mixins.

Oh, no no no... 

If you think that [your example code](https://github.com/jcavalieri/yard_issue/blob/39430b9d7e71eb8d8a304ec771bc2b569e65ea8b/lib/app.rb#L7) is including a mixin only for that class, you're unfortunately very mistaken there. The two versions of the code provided (`Hash.include` vs opening the class) are exactly equivalent. You can test this by simply calling `{}.respond_to?(:select_keys)` anywhere outside that Reproduced class.

As an example: https://repl.it/join/bztzrasb-lsegal

I think what you're looking for is actually [Refinements](https://docs.ruby-lang.org/en/2.4.0/syntax/refinements_rdoc.html), which is fairly different syntactically but would likely work similar to your expectations. That said, Refinements isn't exactly a popular feature-- realistically the more accepted and generic (read: cross-language) convention would be to simply use [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) to filter via an instance of a specialized _class_ that performs said filtering. It requires a rewrite of the way you have it structured, specifically your `Ext::Core::Hash::Filtering` would have to operate on a provided object rather than `self`, since `self` requires it to be included directly onto the Hash class. This is the example of composition:

```ruby
class Filterizer
  def initialize(object) @object = object end
  def select_keys(*args) @object.select { |k, _| args.include?(k) } end
end

class Reproduced
  def do_something(my_hash)
    Filterizer.new(my_hash).select_keys('a', 'b')
  end
end
```

So in short, YARD is interpreting the code correctly here.

**Comment by lsegal** [comment]
PS: I love refinements and am sorry that I called them unpopular. If you're interested in helping refinements see more light of day, here's the refinements version:

```ruby
module Filterable
  refine Hash do
    include Ext::Core::Hash::Filtering
  end
end

HASH_OBJECT = {'foo' => 'one', bar: 'two', baz: 'three'}

# It's filterable in here!
module MyBaseModule
  using Filterable

  p HASH_OBJECT.select_keys('foo', 'bar')
end

# It's not filterable out here!
p HASH_OBJECT.select_keys('foo', 'bar')
```

https://repl.it/join/doevdzsk-lsegal

**Comment by jcavalieri** [comment]
Wow, I was mistaken! Thanks for clarifying that!

And I forgot about refinements. That is does solve what we want to do.

Thanks @lsegal , I'll close this issue.

**Comment by esotericpig** [comment]
I have this same issue when prepending to core:

```Ruby
Module.prepend AttrBool::Ext
```

The nice thing about this is that I don't have to worry about what `Module` is, just that it has the method `prepend` (duck typing), but if I'm forced to use the solution by re-opening the class and adding `@api private` then I lose this and have to specify `class Module`. IMO, I don't think documentation should force me to code in a different way.

I think it'd be nice to have some option on the command-line (`.yardopts`) to fix this.
